### PR TITLE
HHH-17789 fix the blob/clob/nclob not free caused hana memory issue

### PR DESCRIPTION
**Pull request by gaogaodepanpan**
https://hibernate.atlassian.net/browse/HHH-17789

### CODE REVIEW

**Review by songyi10011001 — APPROVED** [review]

**Review by Paul168888 — APPROVED** [review]

**Review by mbladel — APPROVED** [review]
Apart from some code-style changes, this looks reasonable to me.

File: hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java [hunk 1]
```diff
@@ -1529,25 +1529,33 @@ protected void doBind(CallableStatement st, X value, String name, WrapperOptions
...
-					}
-					Blob blob = new MaterializedBlob( DataHelper.extractBytes( rsBlob.getBinaryStream() ) );
-					return javaType.wrap( blob, options );
+					return extract( rs.getBlob( paramIndex ), options );
 				}
 
 				@Override
 				protected X doExtract(CallableStatement statement, int index, WrapperOptions options) throws SQLException {
-					return javaType.wrap( statement.getBlob( index ), options );
+					return extract(statement.getBlob( index ), options );
```
[the remark this answers is not in the document]
Comment (gaogaodepanpan) [reply hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java]: above format updated, thx for your review

File: hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java [hunk 2]
```diff
@@ -1529,25 +1529,33 @@ protected void doBind(CallableStatement st, X value, String name, WrapperOptions
...
 				}
 
 				@Override
 				protected X doExtract(CallableStatement statement, String name, WrapperOptions options) throws SQLException {
-					return javaType.wrap( statement.getBlob( name ), options );
+					return extract(statement.getBlob( name ), options );
```
[the remark this answers is not in the document]
Comment (gaogaodepanpan) [reply hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java]: above format updated, thx for your review

File: hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java [hunk 3]
```diff
@@ -1633,22 +1655,31 @@ protected X doExtract(ResultSet rs, int paramIndex, WrapperOptions options) thro
 					else {
 						rsClob = rs.getClob( paramIndex );
 					}
-
-					if ( rsClob == null || rsClob.length() < HANAClobJdbcType.this.maxLobPrefetchSize ) {
-						return javaType.wrap( rsClob, options );
-					}
-					Clob clob = new MaterializedNClob( DataHelper.extractString( rsClob ) );
-					return javaType.wrap( clob, options );
+					return extract(rsClob, options);
```
[the remark this answers is not in the document]
Comment (gaogaodepanpan) [reply hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java]: above format updated, thx for your review

File: hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java [hunk 4]
```diff
@@ -1633,22 +1655,31 @@ protected X doExtract(ResultSet rs, int paramIndex, WrapperOptions options) thro
...
 				protected X doExtract(CallableStatement statement, int index, WrapperOptions options) throws SQLException {
-					return javaType.wrap( statement.getClob( index ), options );
+					Clob rsClob;
+					if ( HANAClobJdbcType.this.useUnicodeStringTypes ) {
+						rsClob = statement.getNClob( index );
+					}
+					else {
+						rsClob = statement.getClob( index );
+					}
+					return extract(rsClob, options);
```
[the remark this answers is not in the document]
Comment (gaogaodepanpan) [reply hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java]: above format updated, thx for your review

File: hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java [hunk 5]
```diff
@@ -1633,22 +1655,31 @@ protected X doExtract(ResultSet rs, int paramIndex, WrapperOptions options) thro
...
 				protected X doExtract(CallableStatement statement, String name, WrapperOptions options) throws SQLException {
-					return javaType.wrap( statement.getClob( name ), options );
+					Clob rsClob;
+					if ( HANAClobJdbcType.this.useUnicodeStringTypes ) {
+						rsClob = statement.getNClob( name );
+					}
+					else {
+						rsClob = statement.getClob( name );
+					}
+					return extract(rsClob, options);
```
[the remark this answers is not in the document]
Comment (gaogaodepanpan) [reply hibernate-core/src/main/java/org/hibernate/dialect/AbstractHANADialect.java]: above format updated, thx for your review

**Review by beikov — APPROVED** [review]
LGTm

**Review by mbladel — APPROVED** [review]
Thanks @beikov. Last thing @gaogaodepanpan, please squash the code-style commit so that we have a single one and we can merge this, thanks!

**Comment by gaogaodepanpan** [comment]
> Thanks @beikov. Last thing @gaogaodepanpan, please squash the code-style commit so that we have a single one and we can merge this, thanks!

Hi @mbladel, It seems this PR is ready to merge, you will help to merge this to main right?

**Pull request merged**